### ISSUE TITLE
"Trying to get property of non-object" in dompdf/src/Css/Stylesheet.php(1032): Dompdf\Css\Style->length_in_pt(Array)

### ISSUE THREAD

**Issue by MatthiasKuehneEllerhold** [dompdf/dompdf#1406]
We're experiencing this error while converting HTML with CSS to PDF:

```
Trying to get property of non-object in /xxx/vendor/dompdf/dompdf/src/Css/Style.php(584)
Stack trace:
#0 /xxx/vendor/dompdf/dompdf/src/Css/Stylesheet.php(1032): Dompdf\Css\Style->length_in_pt(Array)
#1 /xxx/vendor/dompdf/dompdf/src/Dompdf.php(711): Dompdf\Css\Stylesheet->apply_styles(Object(Dompdf\Frame\FrameTree))
#2 /xxx/PdfCreator.php: Dompdf\Dompdf->render()
```

Relevant software packages in use via composer:
```
dompdf/dompdf      v0.8.0
phenx/php-font-lib  0.5
phenx/php-svg-lib  v0.2
PHP                 7.1.3
```

This issue may be similar to https://github.com/dompdf/dompdf/issues/1008 because we're heavily using "rem" units in our CSS. The resulting PDF seems fine though.

**Comment by bsweeney** [comment]
Are you able to isolate the HTML+CSS that causes the issue? I'm not sure why there would be available style on the root object. Also, what do you think of using the `$ref_size` value the root style doesn't exist rather than skipping the calculation?

**Comment by MatthiasKuehneEllerhold** [comment]
I've managed to use our PDF templates to create a small test case:
```
<!DOCTYPE html>
<html>
    <head>
        <style type="text/css">
            @page {
                margin: 15rem 6rem 12rem;
            }

            html {
                font-size: 10px;
            }

            body {
                position: relative;
                font-family: Helvetica, Arial, sans-serif;
                font-size: 1.2rem;
                color: cmyk(0, 0, 0, 0.89);
            }
        </style>
    </head>
    <body>This is a test!</body>
</html>
```

**Comment by bsweeney** [comment]
To summarize, the problem is with the `rem` units in the `@page` declaration. At the very least this looks to be directly related to the fix for the referenced issue since this didn't appear to happen in the same way prior to 0.8.0.

Since the source of the warning is in the `@page` declaration I believe you should be able to use `em` units instead of `rem` since the size reference will be the same in either case.

**Comment by MatthiasKuehneEllerhold** [comment]
I've updated the PR to reflect your comment, but I'm not sure I've done it correctly. I'd have to know if the element is `@page` in order to switch to the em calculation. How do I do that in the `length_in_pt()` method?
See: https://github.com/dompdf/dompdf/pull/1408/files

**Comment by bsweeney** [comment]
Let me look at your code and I'll get back to you.

**Comment by bsweeney** [comment]
Our problem may be deeper than I initially thought. Because we're calling `length_in_pt()` before the page element has had styles applied, the base font size will be the default. That means calculated dimensions could be based on an incorrect value. For example, in this case the font size is also set in the stylesheet and so using the default is incorrect.

I'm going to have to analyze this a bit more. We may need to move some of the logic in this section so that the page style can be set. Still, your fix is probably necessary to handle edge cases more gracefully until we can figure out how to address the problem more thoroughly.

**Comment by MatthiasKuehneEllerhold** [comment]
@bsweeney What do you want to do with this issue? Close it? Leave it open for further investigation? 

Secondly it would be really nice if you could release a bugfix release soon with the fixes from #1407 and #1408...

**Comment by bsweeney** [comment]
I usually leave them open until I merge to master (github will then close any that are addressed by commits). I hope to tag a release soon. Just a few more issues to review.
